{"original_position": 85, "diff_hunk": "@@ -373,6 +393,52 @@ def save(self, *args, **kwargs):\n         self.implementation_hash = self.compute_implementation_hash()\n         super().save(update_fields=['implementation_hash'])\n \n+    def validate_arguments(self, arguments, old_arguments=None):\n+        \"\"\"\n+        Test if `arguments` follows all action-specific rules.\n+\n+        If there is a previous version of the arguments, they should\n+        be passed as `old_arguments`. Some validation rules depend on\n+        the history of the arguments.\n+\n+        Raises `ValidationError` if any rules are violated.\n+        \"\"\"\n+        if self.name == 'preference-experiment':\n+            # Make a default dict that always returns a default dict\n+            def default():\n+                return defaultdict(default)\n+            errors = default()\n+\n+            # Feature branch slugs should be unique within an experiment.\n+            branch_slugs = set()\n+            branch_values = set()\n+            for i, branch in enumerate(arguments['branches']):\n+                if branch['slug'] in branch_slugs:\n+                    msg = self.errors['duplicate_branch_slug']\n+                    errors['branches'][i]['slug'] = msg\n+\n+                if branch['value'] in branch_values:\n+                    msg = self.errors['duplicate_branch_value']\n+                    errors['branches'][i]['value'] = msg\n+\n+                branch_slugs.add(branch['slug'])\n+                branch_values.add(branch['value'])\n+\n+            # Experiment slugs should be unique.\n+            experiment_recipes = Recipe.objects.filter(latest_revision__action=self)\n+            experiment_slugs = set()", "body_text": "We never add the slugs that we've seen to this set, so this doesn't actually validate for conflicting slugs. I'm pretty sure I tested this on my first pass and it worked, but I can't explain why it might've worked then.", "pull_request_url": "https://api.github.com/repos/mozilla/normandy/pulls/651", "url": "https://api.github.com/repos/mozilla/normandy/pulls/comments/109980529", "created_at": "2017-04-05T17:39:36Z", "author_association": "MEMBER", "body": "We never add the slugs that we've seen to this set, so this doesn't actually validate for conflicting slugs. I'm pretty sure I tested this on my first pass and it worked, but I can't explain why it might've worked then.", "updated_at": "2017-04-05T19:18:07Z", "html_url": "https://github.com/mozilla/normandy/pull/651#discussion_r109980529", "pull_request_review_id": 31118558, "_links": {"self": {"href": "https://api.github.com/repos/mozilla/normandy/pulls/comments/109980529"}, "html": {"href": "https://github.com/mozilla/normandy/pull/651#discussion_r109980529"}, "pull_request": {"href": "https://api.github.com/repos/mozilla/normandy/pulls/651"}}, "commit_id": "4ed398312e1e920bb331944c0d17a5a6bfe0f7cb", "user": {"following_url": "https://api.github.com/users/Osmose/following{/other_user}", "events_url": "https://api.github.com/users/Osmose/events{/privacy}", "organizations_url": "https://api.github.com/users/Osmose/orgs", "url": "https://api.github.com/users/Osmose", "gists_url": "https://api.github.com/users/Osmose/gists{/gist_id}", "html_url": "https://github.com/Osmose", "subscriptions_url": "https://api.github.com/users/Osmose/subscriptions", "avatar_url": "https://avatars1.githubusercontent.com/u/193106?v=4", "repos_url": "https://api.github.com/users/Osmose/repos", "received_events_url": "https://api.github.com/users/Osmose/received_events", "gravatar_id": "", "starred_url": "https://api.github.com/users/Osmose/starred{/owner}{/repo}", "site_admin": false, "login": "Osmose", "type": "User", "id": 193106, "followers_url": "https://api.github.com/users/Osmose/followers"}, "position": null, "path": "recipe-server/normandy/recipes/models.py", "body_html": "<p>We never add the slugs that we've seen to this set, so this doesn't actually validate for conflicting slugs. I'm pretty sure I tested this on my first pass and it worked, but I can't explain why it might've worked then.</p>", "original_commit_id": "3156262a8a71d5325f513b187bfe89c032f2490d", "id": 109980529}